modules/dnn/src/layers/fully_connected_layer.cpp

```diff
@@ -415,6 +422,80 @@ class FullyConnectedLayerImpl CV_FINAL : public InnerProductLayer
         }
     }
 
+#ifdef HAVE_CUDA
+    void forwardCUDA(
+        std::vector<cv::Ptr<BackendWrapper>>& inputs,
+        std::vector<cv::Ptr<BackendWrapper>>& outputs,
+        csl::Workspace& workspace
+    )
+    {
+        CV_UNUSED(workspace);
+
+        for (std::size_t i = 0; i < inputs.size(); i++)
+        {
+            auto input_wrapper = inputs[i].dynamicCast<CUDABackendWrapperFP32>();
+            auto output_wrapper = outputs[i].dynamicCast<CUDABackendWrapperFP32>();
+
+            auto input = input_wrapper->getView();
+            auto output = output_wrapper->getSpan();
+
+            auto actual_dims = input_wrapper->getShape().size();
+            CV_Assert(csl::tensor_utils::get_effective_rank(input) <= actual_dims);
```

Comment:
You should consider locating `get_effective_rank` in the same namespace as the view type so that you can invoke it like `get_effective_rank(input)` and let argument dependent lookup do the right thing.  That is, if `get_effective_rank` is really a part of the API of a view then it should be in the same namespace as the view so ADL will work on it in the normal way.

Reply:
Should I remove the `tensor_ops` namespace? It would reduce some bloat in the layer code.

It will mostly have a truckload of functions (all activations, convolution, pooling and whatnot).

Reply:
I don't have deep opinions about how many namespaces should be used.  My general preference is for fewer namespaces. You can set it up however you think is best.